pennylane/ops/functions/generator.py

```diff
@@ -0,0 +1,175 @@
+    """Return the generator as type :class:`~.Hermitian`,
+    :class:`~.SparseHamiltonian`, or :class:`~.Hamiltonian`,
+    as provided by the original gate.
+    """
+    if isinstance(gen, (qml.Hermitian, qml.SparseHamiltonian)):
+        if not op.inverse:
+            return gen
+
+        param = gen.parameters[0]
+        wires = gen.wires
+
+        return gen.__class__(-param, wires=wires)
+
+    if op.inverse:
+        gen = -1.0 * gen
+
+    return gen
+
+
+def _generator_hamiltonian(gen, op):
+    """Return the generator as type :class:`~.Hamiltonian`."""
+    wires = op.wires
+
+    if isinstance(gen, qml.Hamiltonian):
+        H = gen
+
+    elif isinstance(gen, (qml.Hermitian, qml.SparseHamiltonian)):
+
+        if isinstance(gen, qml.Hermitian):
+            mat = gen.parameters[0]
+
+        elif isinstance(gen, qml.SparseHamiltonian):
+            mat = gen.parameters[0].toarray()
+
+        coeffs, obs = qml.utils.decompose_hamiltonian(mat, wire_order=wires, hide_identity=True)
+        H = qml.Hamiltonian(coeffs, obs)
+
+    elif isinstance(gen, qml.operation.Observable):
+        H = 1.0 * gen
+
+    if op.inverse:
+        H = -1.0 * H
+
+    return H
+
+
+def _generator_prefactor(gen, op):
+    r"""Return the generator as ```(prefactor, obs)`` representing
+    :math:`G=p \hat{O}`, where
+
+    - prefactor :math:`p` is a float
+    - observable `\hat{O}` is one of :class:`~.Hermitian`,
+      :class:`~.SparseHamiltonian`, or a tensor product
+      of Pauli words.
+    """
+    if isinstance(gen, (qml.Hermitian, qml.SparseHamiltonian)):
+        obs = gen
+        prefactor = 1.0
+
+    elif isinstance(gen, qml.operation.Observable):
+        # convert to a qml.Hamiltonian
+        gen = 1.0 * gen
+
+        if len(gen.ops) == 1:
+            # case where the Hamiltonian is a single Pauli word
+            obs = gen.ops[0]
+            prefactor = gen.coeffs[0]
+        else:
+            obs = gen
+            prefactor = 1.0
+
+    if op.inverse:
+        prefactor *= -1.0
+
+    return prefactor, obs
+
+
+@qml.op_transform
+def generator(op, format="prefactor"):
+    r"""Returns the generator of operation.
+
+    Args:
+        op (.Operator or Callable): A single operator, or a function that
+            applies a single quantum operation.
+        format (str): The format to return the generator in. Must be one of ``'prefactor'``,
+            ``'observable'``, or ``'hamiltonian'``. See below for more details.
+
+    Returns:
+        .Observable or tuple[float, .Observable]: The returned generator, with format/type
+        dependent on the ``format`` argument.
+
+        * ``"prefactor"``: Return the generator as ```(prefactor, obs)`` (representing
+          :math:`G=p \hat{O}`), where
+
+          - prefactor :math:`p` is a float
+          - observable `\hat{O}` is one of :class:`~.Hermitian`,
+            :class:`~.SparseHamiltonian`, or a tensor product
+            of Pauli words.
```

Comment:
This decomposition is not unique, right? Should we say something about how `qml.generator` decides what is the prefactor and what is the `obs`?

Reply:
Good point! I'll add something. Basically: the prefactor is always `±1.0`, unless the generator is a single Pauli word of the form `p * Word`, in which case the prefactor is `p`.